cipher/transposition/transposition_test.go

```diff
@@ -93,11 +92,54 @@ func TestEncryptDecrypt(t *testing.T) {
 	if errDecrypt != nil {
 		t.Error(errDecrypt)
 	}
-	if strings.Contains(decrypt, text) == false {
-		t.Error("The string was not decrypted correctly")
+	if decrypt != text {
+		t.Errorf("The string was not decrypted correctly %q %q", decrypt, text)
 	}
 	decrypt, _ = Decrypt([]rune(encrypt), key2)
-	if strings.Contains(decrypt, text) == true {
-		t.Error("The string was decrypted with a different key")
+	if decrypt == text {
+		t.Errorf("The string was decrypted with a different key: %q %q", decrypt, text)
 	}
 }
+
+func FuzzTransposition(f *testing.F) {
+	for _, transpositionTestInput := range getTexts() {
+		f.Add(transpositionTestInput)
+	}
+	f.Fuzz(func(t *testing.T, input string) {
+		keyword := getRandomString()
+		encrypted, err := Encrypt([]rune(input), keyword)
+		switch {
+		case err == nil:
+		case errors.Is(err, ErrKeyMissing),
+			errors.Is(err, ErrNoTextToEncrypt):
+			return
+		default:
+			t.Fatalf("unexpected error when encrypting string %q: %v", input, err)
+		}
+		decrypted, err := Decrypt([]rune(encrypted), keyword)
+		switch {
+		case err == nil:
+		case errors.Is(err, ErrKeyMissing),
+			errors.Is(err, ErrNoTextToEncrypt):
+			return
+		default:
+			t.Fatalf("unexpected error when decrypting string %q: %v", encrypted, err)
+		}
+
+		if !equals([]rune(decrypted), []rune(input)) {
+			t.Fatalf("expected: %+v, got: %+v", []rune(input), []rune(decrypted))
+		}
+	})
+}
+
+func equals(r, s []rune) bool {
+	if len(r) != len(s) {
+		return false
+	}
+	for i := range r {
+		if r[i] != s[i] {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
Seems that you have changed the implementation. Isn't `string(r) == string(s)` working?

Reply:
Unfortunately it doesn't work when it compares an input string that are not utf8 valid, i.e. containing special characters.

With this check:
```go
if string(decrypted) != string(input) {
	t.Logf("comparing %q with %q as strings", string(input), string(decrypted))
	t.Logf("comparing %q with %q as []rune: equals? %t", []rune(input), []rune(decrypted), equals([]rune(input), []rune(decrypted)))
	t.Logf("checking utf-8 compliance of input: %t; and decrypted: %t", utf8.ValidString(input), utf8.ValidString(decrypted))
	t.Fatalf("expected: %+v, got: %+v", []rune(input), []rune(decrypted))
}
```

I can see that the decrypted not utd8 valid string is encoded in an utf8 valid []rune when the string is assigned to `Decrypt` and this makes the comparison as string at the end failing; even though the rune representation is the same.

```bash
$ go test ./cipher/transposition -fuzz=^FuzzTransposition$
fuzz: elapsed: 0s, gathering baseline coverage: 0/464 completed
failure while testing seed corpus entry: FuzzTransposition/00d38231075839cae8d63ba3ea8aff49bfcfd394f690a57e1cd2923ba0f4d7e3
fuzz: elapsed: 0s, gathering baseline coverage: 2/464 completed
--- FAIL: FuzzTransposition (0.04s)
    --- FAIL: FuzzTransposition (0.00s)
        transposition_test.go:131: comparing "\xcea9" with "�a9" as strings
        transposition_test.go:132: comparing ['�' 'a' '9'] with ['�' 'a' '9'] as []rune
        transposition_test.go:133: checking utf-8 compliance of input: false; and decrypted: true
        transposition_test.go:134: expected: [65533 97 57], got: [65533 97 57]
    
FAIL
exit status 1
FAIL    github.com/TheAlgorithms/Go/cipher/transposition        0.047s
```

So I can present you these alternatives:
1) remove and equals func I created and use `decrypted != string([]rune(input))` to compare input and result
2) before encrypting a string filter out all non utf8 valid strings and return error in case the input is not utf8 valid
3) leave the check with the equals (or reflect.DeepEqual if you prefer) as the underlying []rune representation for both input and decrypted string is the same.

What do you think is the best solution for you?

Reply:
Thanks for the comprehensive information. I have a few thoughts, let me know what you think 😄 


The solution I see is none of those, but rather fixing the main problem of conversion in the function signatures. This is not always possible in real world but in this case nothing in our project makes any backwards compatibility promises therefore you can break the api contracts (for example function signatures - feel free to change them if they are indeed the problem rather than overcoming issues in its usage).

```Go
func Encrypt(text []rune, keyWord string) ([]rune, error) {
```
instead of 
```Go
func Encrypt(text []rune, keyWord string) (string, error) {
```

and 
```Go
func Decrypt(text []rune, keyWord string) ([]rune, error) {
```
instead of 
```Go
func Decrypt(text []rune, keyWord string) ([]rune, error) {
```

Once you have that, you can either use `reflect.DeepEqual` or cast them to strings and compare like so `string(decrypt) != string(input)` or something along those lines.

Reply:
You're welcome! And thanks for your explanation and suggestion as well: in this situation I'm ok to modify the signature of the function and adapt the tests accordingly. I updated the PR, let me know if you agree with the kind of change I did. Thanks again!